File: 3-UML Diagrams/Library Information System/library/Library.java
package library;

import items.AVMaterial;
import items.Book;
import items.Item;
import items.LoanableItem;
import items.Magazine;
import items.ReferenceBook;
import patrons.Patron;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Library class.
 *
 * @author Ziang Lu
 */
public class Library {

    /**
     * Next library card number to assign.
     */
    private static int patronCardNum = 0;

    /**
     * Items by title.
     */
    private Map<String, List<Item>> itemsByTitle;
    /**
     * Patrons by library card number.
     */
    private Map<Integer, Patron> patrons;
    /**
     * Titles of the pending requests.
     */
    private List<String> pendingRequests;

    /**
     * Default constructor.
     */
    public Library() {
        itemsByTitle = new HashMap<>();
        patrons = new HashMap<>();
    }

    /**
     * Private helper method the given item.
     * @param newItem item to add
     */
    private void addItem(Item newItem) {
        String title = newItem.getTitle();
        List<Item> items = itemsByTitle.getOrDefault(title, new ArrayList<>());
        items.add(newItem);
        itemsByTitle.put(title, items);
    }

    /**
     * Adds a new book.
     * @param title title of the book
     * @param value value of the book
     * @param bestSeller whether the book to add is a best-seller
     */
    public void addBook(String title, double value, boolean bestSeller) {
        Book newBook = new Book(title, value, bestSeller);
        addItem(newBook);
    }

    /**
     * Adds a new A/V material.
     * @param title title of the A/V material
     * @param value value of the A/V material
     */
    public void addAVMaterial(String title, double value) {
        AVMaterial newAVMaterial = new AVMaterial(title, value);
        addItem(newAVMaterial);
    }

    /**
     * Adds a new reference book.
     * @param title title of the reference book
     */
    public void addReferenceBook(String title) {
        ReferenceBook newReferenceBook = new ReferenceBook(title);
        addItem(newReferenceBook);
    }

    /**
     * Adds a new magazine.
     * @param title title of the magazine
     */
    public void addMagazine(String title) {
        Magazine newMagazine = new Magazine(title);
        addItem(newMagazine);
    }

    /**
     * Registers a new patron.
     * @param name name of the patron
     * @param age age of the patron
     * @param phoneNum phone number of the patron
     * @param addr address of the patron
     * @return newly registered patron
     */
    public Patron registerPatron(String name, int age, long phoneNum, String addr) {
        Patron newPatron = new Patron(patronCardNum, name, age);
        newPatron.setPhoneNum(phoneNum);
        newPatron.setAddr(addr);
        patrons.put(patronCardNum, newPatron);
        ++patronCardNum;
        return newPatron;
    }

    /**
     * Lets the given patron requests an item with the given title.
     * @param title title of the item to request
     * @return requested item if available, null if not
     */
    public LoanableItem itemRequested(String title) {
        if (!itemsByTitle.containsKey(title)) {
            return null;
        }
        List<Item> titledItems = itemsByTitle.get(title);
        for (Item titledItem : titledItems) {
            if (titledItem instanceof LoanableItem) {
                LoanableItem titledLoanable = (LoanableItem) titledItem;
                if (!titledLoanable.isCheckedOut()) {
                    titledLoanable.checkOut();
                    return titledLoanable;
                }
            }
        }
        pendingRequests.add(title);
        return null;
    }

    /**
     * Lets the given item to be renewed.
     * @param item item to renew
     * @return whether the item is successfully renewed
     */
    public boolean itemRenewed(Item item) {
        LoanableItem loanable = (LoanableItem) item;
        // An item can only be renewed once.
        if (loanable.isRenewed()) {
            return false;
        }
        // If there is a pending request for this item, it cannot be renewed.
        String title = item.getTitle();
        for (String pendingRequest : pendingRequests) {
            if (pendingRequest.equals(title)) {
                return false;
            }
        }
        ((LoanableItem) item).renew();
        return true;
    }

    /**
     * Lets the given item to be returned.
     * @param item item to return
     */
    public void itemReturned(Item item) {
        ((LoanableItem) item).checkIn();
        // Remove the corresponding pending request if there is one
        pendingRequests.remove(item.getTitle());
    }

    /**
     * Gets the checked out items by the given patron.
     * @param cardNum library card number of the patron
     * @return checked out items by the patron
     */
    public List<Item> getPatronItems(int cardNum) {
        Patron patron = patrons.get(cardNum);
        if (patron == null) {
            return new ArrayList<>();
        }
        return patron.getMyItems();
    }

    /**
     * Calculates the overdue fine of the given patron.
     * @param cardNum library card number of the patron
     * @return overdue fine of the patron
     */
    public double calcOverdueFine(int cardNum) {
        Patron patron = patrons.get(cardNum);
        if (patron == null) {
            return 0.0;
        }
        return patron.calcOverdueFine();
    }

}

File: 3-UML Diagrams/Library Information System/Python/library.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Library module.
"""

__author__ = 'Ziang Lu'

from typing import Iterable

from items import AVMaterial, Book, Item, LoanableItem, Magazine, ReferenceBook
from patron import Patron


class Library(object):
    """
    Library class.
    """
    __slots__ = ['_items_by_title', '_patrons', '_pending_requests']
    _patron_card_num = 0

    def __init__(self):
        """
        Default constructor.
        """
        self._items_by_title = {}
        self._patrons = {}
        self._pending_requests = []

    def _add_item(self, new_item: Item) -> None:
        """
        Private helper method to add the given item.
        :param new_item: item to add
        :return: None
        """
        title = new_item.title
        items = self._items_by_title.get(title, [])
        items.append(new_item)
        self._items_by_title[title] = items

    def add_book(self, title: str, value: float, best_seller: bool) -> None:
        """
        Adds a new book
        :param title: str
        :param value: float
        :param best_seller: bool
        :return: None
        """
        new_book = Book(title, value, best_seller)
        self._add_item(new_book)

    def add_av_material(self, title: str, value: float) -> None:
        """
        Adds a new A/V material.
        :param title: str
        :param value: float
        :return: None
        """
        new_av_material = AVMaterial(title, value)
        self._add_item(new_av_material)

    def add_reference_book(self, title: str) -> None:
        """
        Adds a new reference book.
        :param title: str
        :return: None
        """
        new_reference_book = ReferenceBook(title)
        self._add_item(new_reference_book)

    def add_magazine(self, title: str) -> None:
        """
        Adds a new magazine.
        :param title: str
        :return: None
        """
        new_magazine = Magazine(title)
        self._add_item(new_magazine)

    def register_patron(self, name: str, bday: int, phone_num: int,
                        addr: str) -> Patron:
        """
        Registers a new patron.
        :param name: str
        :param bday: datetime object
        :param phone_num: int
        :param addr: str
        :return: Patron
        """
        new_patron = Patron(self._patron_card_num, name, bday)
        new_patron.phone_num = phone_num
        new_patron.addr = addr
        self._patrons[self._patron_card_num] = new_patron
        self._patron_card_num += 1
        return new_patron

    def item_requested(self, title: str) -> LoanableItem:
        """
        Lets the given patron requests an item with the given title.
        :param title: str
        :return: LoanableItem
        """
        if title not in self._items_by_title:
            return None
        titled_items = self._items_by_title[title]
        for titled_item in titled_items:
            if isinstance(titled_item, LoanableItem):
                if not titled_item.is_checked_out:
                    titled_item.check_out()
                    return titled_item
        self._pending_requests.append(title)
        return None

    def item_renewed(self, item: LoanableItem) -> bool:
        """
        Lets the given item to be renewed.
        :param item: LoanableItem
        :return: bool
        """
        # An item can only be renewed once.
        if item.is_renewed:
            return False
        # If there is a pending request for this item, it cannot be renewed.
        title = item.title
        for pending_request in self._pending_requests:
            if pending_request == title:
                return False
        item.renew()
        return True

    def item_returned(self, item: LoanableItem) -> None:
        """
        Lets the given item to be returned.
        :param item: LoanableItem
        :return: None
        """
        item.check_in()
        # Remove the corresponding pending request if there is one
        self._pending_requests.remove(item.title)

    def get_patron_items(self, card_num: int) -> Iterable[Item]:
        """
        Gets the checked out items by the given patron.
        :param card_num: int
        :return: iterable[Item]
        """
        if card_num not in self._patrons:
            return []
        return self._patrons[card_num].items

    def calc_overdue_fine(self, card_num: int) -> float:
        """
        Calculates the overdue fine of the given patron.
        :param card_num: int
        :return: float
        """
        if card_num not in self._patrons:
            return 0.0
        return self._patrons[card_num].calc_overdue_fine()
